### PR TITLE
8330691: [lworld] C2: assert(gvn().type(n)->is_zero_type()) failed: Should have been scalarized

### PR DESCRIPTION
**Pull request by chhagedorn**
The unique concrete subclass optimization replaces an abstract class with a unique concrete sub class (if there is one). In the test case, `abstractValueClassSingleSubclass` is a field with the abstract value class type `AbstractValueClassSingleSubclass` which has a unique (loaded) concrete value sub class `UniqueValueSubClass`.

In `do_get_xxx()`, we get the declared `ciType` of the field:
https://github.com/openjdk/valhalla/blob/3b89acb102ddb7b4f318edab5c1a6169b28757c0/src/hotspot/share/opto/parse3.cpp#L143

which is `AbstractValueClassSingleSubclass` (i.e. no unique concrete sub class optimization applied). But for the type, we call `make_from_klass()` which does apply the unique concrete sub class optimization:
https://github.com/openjdk/valhalla/blob/3b89acb102ddb7b4f318edab5c1a6169b28757c0/src/hotspot/share/opto/parse3.cpp#L173
https://github.com/openjdk/valhalla/blob/3b89acb102ddb7b4f318edab5c1a6169b28757c0/src/hotspot/share/opto/type.cpp#L3842-L3859

As a result, we create a load that has type `UniqueValueSubClass` while `field_klass` is still abstract. We do not create an `InlineTypeNode` because `is_inlinetype()` returns false for abstract classes:
https://github.com/openjdk/valhalla/blob/3b89acb102ddb7b4f318edab5c1a6169b28757c0/src/hotspot/share/opto/parse3.cpp#L191-L193

Later, we assert that we've missed a scalarization opportunity based on the concrete inline type load.

The fix is straight forward to improve the fetched declared `ciType` in `do_get_xxx()` by applying the unique concrete sub class optimization to match the improved type. We only need to apply it for abstract inline type classes. In other cases, we do not the improved field type. I've added a method to check this accordingly.

Thanks,
Christian

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Issue
 * [JDK-8330691](https://bugs.openjdk.org/browse/JDK-8330691): [lworld] C2: assert(gvn().type(n)-&gt;is_zero_type()) failed: Should have been scalarized (**Bug** - P4)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - Committer) ⚠️ Review applies to [9fa77a26](https://git.openjdk.org/valhalla/pull/1099/files/9fa77a26ee0bbfc13799ccd3f6ce2e59bd7791ec)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/valhalla.git pull/1099/head:pull/1099` \
`$ git checkout pull/1099`

Update a local copy of the PR: \
`$ git checkout pull/1099` \
`$ git pull https://git.openjdk.org/valhalla.git pull/1099/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1099`

View PR using the GUI difftool: \
`$ git pr show -t 1099`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/valhalla/pull/1099.diff">https://git.openjdk.org/valhalla/pull/1099.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/valhalla/pull/1099#issuecomment-2100105723)

### CODE REVIEW

File: test/hotspot/jtreg/compiler/valhalla/inlinetypes/TestLWorld.java [hunk 1]
```diff
@@ -4372,4 +4372,29 @@ public void test160_verifier(RunInfo info) {
...
+
+    value static class UniqueValueSubClass extends AbstractValueClassSingleSubclass {
+        int x = 34;
+    }
+
+    static AbstractValueClassSingleSubclass abstractValueClassSingleSubclass = new UniqueValueSubClass();
+
+    @Test
+    public void testUniqueConcreteValueSubKlass(boolean flag) {
+        // C2 should recognize that even though we do not know whether the underlying inline type of the abstract field
```
[the remark this answers is not in the document]
Comment (chhagedorn) [reply test/hotspot/jtreg/compiler/valhalla/inlinetypes/TestLWorld.java]: Thanks for the hint, pushed an update.

**Comment by chhagedorn** [comment]
Thanks Tobias for your review! I've merged in your changes and I started testing again.

**Comment by TobiHartmann** [comment]
Thanks, still good!

**Comment by chhagedorn** [comment]
Thanks Tobias for your review! Testing showed no new failures.

/integrate

**Pull request closed without merging**